File: core/src/com/broll/gainea/server/core/objects/MonsterBehavior.java
package com.broll.gainea.server.core.objects;

import com.broll.gainea.misc.RandomUtils;
import com.broll.gainea.server.core.GameContainer;
import com.broll.gainea.server.core.map.Location;
import com.broll.gainea.server.core.map.Ship;
import com.broll.gainea.server.core.utils.LocationUtils;
import com.broll.gainea.server.core.utils.UnitControl;
import com.google.common.collect.Lists;

import java.util.function.BiConsumer;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public enum MonsterBehavior {
    RESIDENT((game, monster) -> {
    }),
    RANDOM((game, monster) -> {
        Location target = RandomUtils.pickRandom(getPossibleTargets(monster).collect(Collectors.toList()));
        if (target != null) {
            UnitControl.conquer(game, Lists.newArrayList(monster), target);
        }
    }),
    AGGRESSIVE((game, monster) -> {
        Location target = RandomUtils.pickRandom(getPossibleTargets(monster).filter(it -> !LocationUtils.noControlledUnits(it)).collect(Collectors.toList()));
        if (target != null) {
            UnitControl.conquer(game, Lists.newArrayList(monster), target);
        } else {
            RANDOM.doAction(game, monster);
        }
    }),
    FLEEING((game, monster) -> {
        Location target = RandomUtils.pickRandom(getPossibleTargets(monster).filter(Location::isFree).collect(Collectors.toList()));
        if (target != null) {
            UnitControl.move(game, monster, target);
        }
    });

    private BiConsumer<GameContainer, Monster> action;

    MonsterBehavior(BiConsumer<GameContainer, Monster> action) {
        this.action = action;
    }

    public void doAction(GameContainer game, Monster monster) {
        this.action.accept(game, monster);
    }

    private static Stream<Location> getPossibleTargets(Monster monster) {
        return monster.getLocation().getConnectedLocations().stream().filter(it -> !(it instanceof Ship));
    }

}

File: server/src/main/com/broll/gainea/server/core/objects/monster/MonsterBehavior.kt
package com.broll.gainea.server.core.objects.monster

import com.broll.gainea.server.core.Game
import com.broll.gainea.server.core.map.Location
import com.broll.gainea.server.core.utils.UnitControl.conquer
import com.broll.gainea.server.core.utils.UnitControl.move


private fun Monster.getPossibleTargets() = moveTargets
private fun Location.getEnemyStrength() = units.sumOf { it.battleStrength }

enum class MonsterBehavior(val label: String, private val action: (Game, Monster) -> kotlin.Unit) {
    RESIDENT("Sesshaft", { _, _ -> }),
    RANDOM("Wild", { game, monster ->
        monster.getPossibleTargets().randomOrNull()?.let {
            game.conquer(listOf(monster), it)
        }
    }),
    AGGRESSIVE("Aggressiv",
            { game, monster ->
                val target = monster.getPossibleTargets().filter {
                    !it.free && monster.battleStrength > it.getEnemyStrength()
                }.randomOrNull()
                if (target != null) {
                    game.conquer(listOf(monster), target)
                } else {
                    RANDOM.action(game, monster)
                }
            }),
    FLEEING("Scheu",
            { game, monster ->
                monster.getPossibleTargets().filter { it.free }.randomOrNull()?.let {
                    game.move(monster, it)
                }
            }),
    FRIENDLY("Freundlich", { game, monster ->
        monster.getPossibleTargets().randomOrNull()?.let {
            game.move(monster, it)
        }
    });

    fun doAction(game: Game, monster: Monster) = action(game, monster)


}
